com.ibm.streamsx.topology/opt/python/include/splpy_op.h

```diff
@@ -166,9 +209,132 @@ class SplpyOp {
+
+            if (!pickledCallable) {
+              // The callable cannot be pickled.  Throw an exception that
+              // shuts down the operator, unless it is supressed.
+              // If it is suppressed, this operator will continue to run, but
+              // with no checkpointing enabled.
+              if (PyErr_Occurred()) {
+                SplpyExceptionInfo exceptionInfo = SplpyExceptionInfo::pythonError("setup");
+                if (exceptionInfo.pyValue_) {
+                  SPL::rstring text;
+                  // note pyRStringFromPyObject returns zero on success
+                  if (pyRStringFromPyObject(text, exceptionInfo.pyValue_) == 0) {
+                    msg << " because of python error " << text;
+                  }
+
+                  SPLAPPTRC(L_WARN, msg.str(), "python");
+                  // Offer the exception to the operator, so the operator
+                  // can suppress it.
+                  if (exceptionRaised(exceptionInfo) == 0) {
+                    throw exceptionInfo.exception();
+                  }
+                  exceptionInfo.clear();
+                  // The exception was suppressed.  Continue with
+                  // pickledCallable == NULL, which will cause a do-nothing
+                  // state handler to be created.
+                  // Effectively, checkpointing will not be enabled for
+                  // this operator even though it is stateful.
+                  SPLAPPTRC(L_WARN, "Proceeding with no checkpointing for the " << op()->getContext().getName() << " operator", "python");
+                }
+              }
+              else {
+                // This is probably unreachable.  PyObject_CallObject
+                // returned NULL, but there was no python exception.
+                throw SplpyGeneral::generalException("setup", msg.str());
+              }
+            }
+          }
+          assert(!stateHandler);
+          if (stateful && pickledCallable) {
+            SPLAPPTRC(L_DEBUG, "Creating functional state handler", "python");
+            // pickledCallable reference stolen here.
+            stateHandler = new SplpyOpStateHandlerImpl(this, pickledCallable);
+            stateHandlerMutex = stateHandler->getMutex();
+          }
+          else {
+            SPLAPPTRC(L_DEBUG, "Creating nonfunctional state handler", "python");
+            stateHandler = new SplpyOpStateHandler;
+            stateHandlerMutex = NULL;
+          }
+          SPLAPPTRC(L_DEBUG, "registerStateHandler", "python");
+          op()->getContext().registerStateHandler(*stateHandler);
+        }
+      }
+
+      class RealAutoLock {
+      public:
+        RealAutoLock(SplpyOp * op) : op_(op) {
+          locked_ = (op->stateHandlerMutex != NULL);
+          if (locked_){
+            op->stateHandlerMutex->lock();
+          }
+        }
+        ~RealAutoLock() {
+          unlock();
+        }
+        void unlock() {
+          if (locked_) {
```

Comment:
Shouldn't `unlocked()` reset `locked_` to false?

Reply:
Yes, good catch.  I have updated this.